pkg/documentlog/store/sql.go

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (C) 2020. Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package store
+
+import (
+	"bytes"
+	"errors"
+	"github.com/golang-migrate/migrate/v4"
+	"github.com/golang-migrate/migrate/v4/database/sqlite3"
+	bindata "github.com/golang-migrate/migrate/v4/source/go_bindata"
+	"github.com/jinzhu/gorm"
+	logging "github.com/nuts-foundation/nuts-network/logging"
+	"github.com/nuts-foundation/nuts-network/migrations"
+	"sync"
+
+	// import needed to enable the sqlite dialect for gorm
+	_ "github.com/jinzhu/gorm/dialects/sqlite"
+	"github.com/nuts-foundation/nuts-network/pkg/model"
+	"io"
+	"time"
+)
+
+const documentTable = "document"
+
+var documentColumns = []string{"hash", "type", "timestamp", "consistency_hash", "LENGTH(contents) as size"}
+
+func NewSQLDocumentStore(db *gorm.DB) (DocumentStore, error) {
+	docStore := &sqlDocumentStore{db: db, writeMutex: &sync.Mutex{}}
+	if err := docStore.migrate(); err != nil {
+		return nil, err
+	}
+	return docStore, nil
+}
+
+type sqlDocumentStore struct {
+	db         *gorm.DB
+	writeMutex *sync.Mutex
+}
+
+type sqlDocument struct {
+	Hash            string `gorm:"type:char(40);PRIMARY_KEY"`
+	ConsistencyHash string `gorm:"column:consistency_hash;type:char(40);PRIMARY_KEY"`
+	Type            string `gorm:"type:varchar(100)"`
+	Timestamp       int64  `gorm:"type:bigint"`
+	Size            int    `gorm:"-"`
+	Contents        []byte `gorm:"type:blob"`
+}
+
+func (d sqlDocument) storedDocument() StoredDocument {
+	hash, _ := model.ParseHash(d.Hash)
+	consistencyHash, _ := model.ParseHash(d.ConsistencyHash)
+	return StoredDocument{
+		HasContents:     d.Size > 0,
+		ConsistencyHash: consistencyHash,
+		Document: model.Document{
+			Hash:      hash,
+			Type:      d.Type,
+			Timestamp: time.Unix(0, d.Timestamp).UTC(),
+		},
+	}
+}
+
+func (s sqlDocumentStore) Add(document model.Document) (model.Hash, error) {
```

Comment:
I've thought about this for a long time, too. A few considerations that come to mind:

- A relational database isn't the optimal choice to store the document log in, since it's they're documents.
- SQLite lacks the features to optimize this on the database side (e.g. using stored procedures)
- Not accepting older events older than a certain window (a day) makes the potential problem much smaller, but the publishing side wouldn't get any feedback why other nodes aren't processing its hashes (in the current version of the protocol). That'd make diagnosing errors quite hard. I'd rather fix this when we introduce the consensus algorithm since I'd expect it to fix this problem, too. This isn't going to be a problem until we have lots of data, anyways.

So for now I'll optimize it by just recalculating records which's timestamps are larger than the document being added.